File: day20_part1_2021.dart
import 'dart:io';

void main() {
  var lines = File('input.txt').readAsLinesSync();
  var algorithm = lines[0];
  var image = lines.sublist(2).map((line) => line.runes.toList()).toList();
  image = enhanceImage(image, algorithm, 2);
  print(countLitPixels(image));
}

List<List<int>> enhanceImage(List<List<int>> image, String algorithm, int times) {
  for (var i = 0; i < times; i++) {
    image = applyAlgorithm(image, algorithm, i % 2 == 1 && algorithm[0] == '#');
  }
  return image;
}

List<List<int>> applyAlgorithm(List<List<int>> image, String algorithm, bool flip) {
  var enhancedImage = List.generate(image.length + 2, (index) => List.filled(image[0].length + 2, 0));
  for (var i = 0; i < enhancedImage.length; i++) {
    for (var j = 0; j < enhancedImage[i].length; j++) {
      var index = calculateIndex(i - 1, j - 1, image, flip);
      enhancedImage[i][j] = algorithm.codeUnitAt(index);
    }
  }
  return enhancedImage;
}

int calculateIndex(int i, int j, List<List<int>> image, bool flip) {
  var index = 0;
  for (var di = -1; di <= 1; di++) {
    for (var dj = -1; dj <= 1; dj++) {
      index <<= 1;
      if (i + di >= 0 && i + di < image.length && j + dj >= 0 && j + dj < image[0].length) {
        if (image[i + di][j + dj] == '#'.codeUnitAt(0)) {
          index |= 1;
        }
      } else if (flip) {
        index |= 1;
      }
    }
  }
  return index;
}

int countLitPixels(List<List<int>> image) {
  var count = 0;
  for (var row in image) {
    for (var pixel in row) {
      if (pixel == '#'.codeUnitAt(0)) {
        count++;
      }
    }
  }
  return count;
}

File: objectivec/day20_part1_2021.m
#import <Foundation/Foundation.h>

#define ALGORITHM_LENGTH 512

NSString *readFile(NSString *filename) {
    NSString *filePath = [[NSBundle mainBundle] pathForResource:@"input" ofType:@"txt"];
    NSString *fileContents = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
    return fileContents;
}

NSArray<NSString *> *parseInput(NSString *input) {
    NSArray<NSString *> *lines = [input componentsSeparatedByString:@"\n"];
    NSMutableArray<NSString *> *algorithm = [NSMutableArray array];
    NSMutableArray<NSString *> *image = [NSMutableArray array];
    BOOL isAlgorithm = YES;

    for (NSString *line in lines) {
        if ([line isEqualToString:@""]) {
            isAlgorithm = NO;
            continue;
        }
        if (isAlgorithm) {
            [algorithm addObject:line];
        } else {
            [image addObject:line];
        }
    }

    return @[algorithm, image];
}

int getIndexForPixel(NSArray<NSString *> *image, int x, int y, char defaultPixel) {
    int index = 0;
    for (int j = -1; j <= 1; j++) {
        for (int i = -1; i <= 1; i++) {
            index <<= 1;
            if (x + i >= 0 && x + i < image[0].length && y + j >= 0 && y + j < image.count) {
                char pixel = [image[y + j] characterAtIndex:x + i];
                index |= (pixel == '#') ? 1 : 0;
            } else {
                index |= (defaultPixel == '#') ? 1 : 0;
            }
        }
    }
    return index;
}

NSArray<NSString *> *enhanceImage(NSArray<NSString *> *image, NSString *algorithm, char defaultPixel) {
    NSMutableArray<NSString *> *newImage = [NSMutableArray array];
    int width = image[0].length;
    int height = image.count;

    for (int y = -1; y <= height; y++) {
        NSMutableString *row = [NSMutableString string];
        for (int x = -1; x <= width; x++) {
            int index = getIndexForPixel(image, x, y, defaultPixel);
            char newPixel = [algorithm characterAtIndex:index];
            [row appendFormat:@"%c", newPixel];
        }
        [newImage addObject:row];
    }

    return newImage;
}

int countLitPixels(NSArray<NSString *> *image) {
    int count = 0;
    for (NSString *row in image) {
        for (int i = 0; i < row.length; i++) {
            if ([row characterAtIndex:i] == '#') {
                count++;
            }
        }
    }
    return count;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *input = readFile(@"input.txt");
        NSArray<NSString *> *parsedInput = parseInput(input);
        NSArray<NSString *> *algorithmParts = parsedInput[0];
        NSArray<NSString *> *image = parsedInput[1];

        NSString *algorithm = [algorithmParts componentsJoinedByString:@""];

        char defaultPixel = '.';
        image = enhanceImage(image, algorithm, defaultPixel);
        defaultPixel = ([algorithm characterAtIndex:0] == '#') ? '#' : '.';
        image = enhanceImage(image, algorithm, defaultPixel);

        int litPixels = countLitPixels(image);
        printf("Number of lit pixels: %d\n", litPixels);
    }
    return 0;
}